File: src/commands/informations/Account.ts
import {
    ActionRowBuilder,
    ApplicationIntegrationType,
    ButtonBuilder,
    ButtonStyle,
    ContainerBuilder,
    InteractionContextType,
    MessageFlags,
    TextDisplayBuilder,
} from "discord.js";
import { OrionAPIError } from "@/modules/orion-api";
import { Command, CommandCategory, type CommandData, type CommandHandlerContext } from "@/structures/index";

export default class extends Command {
    public static override readonly data: CommandData = {
        name: "account",
        category: CommandCategory.Informations,
        integrationTypes: [ApplicationIntegrationType.GuildInstall, ApplicationIntegrationType.UserInstall],
        contexts: [InteractionContextType.Guild, InteractionContextType.BotDM, InteractionContextType.PrivateChannel],
    };

    public override async handleCommand({ lang, interaction }: CommandHandlerContext) {
        try {
            const user = await this.client.orionAPI.getUser(interaction.user.id);

            const container = new ContainerBuilder()
                .setAccentColor(this.colors.primary.int)
                .addTextDisplayComponents(
                    new TextDisplayBuilder().setContent(
                        `## ${this.emojis.logo} Orion - ${lang.t("card.title.your_account")}`,
                    ),
                )
                .addTextDisplayComponents(
                    new TextDisplayBuilder().setContent(
                        lang.t("card.content.credits", { amount: user.credits.toString() }),
                    ),
                )
                .addActionRowComponents(
                    new ActionRowBuilder<ButtonBuilder>().setComponents(
                        new ButtonBuilder()
                            .setStyle(ButtonStyle.Link)
                            .setLabel(lang.common("dashboard"))
                            .setURL(`${this.config.domainURL}/dashboard`),
                        new ButtonBuilder()
                            .setStyle(ButtonStyle.Link)
                            .setLabel(lang.common("panel"))
                            .setURL(this.config.panelURL),
                    ),
                );

            await interaction.reply({
                components: [container],
                flags: MessageFlags.IsComponentsV2,
                allowedMentions: { parse: [] },
            });
        } catch (err) {
            if (!(err instanceof OrionAPIError) || err.status !== 404) {
                throw err;
            }

            const container = new ContainerBuilder()
                .setAccentColor(this.colors.primary.int)
                .addTextDisplayComponents(new TextDisplayBuilder().setContent(`## ${this.emojis.logo} Orion`))
                .addTextDisplayComponents(new TextDisplayBuilder().setContent(lang.t("no_account.content.guide")))
                .addActionRowComponents(
                    new ActionRowBuilder<ButtonBuilder>().setComponents(
                        new ButtonBuilder()
                            .setStyle(ButtonStyle.Link)
                            .setLabel(lang.t("no_account.buttons.create_account"))
                            .setURL(`${this.config.domainURL}/dashboard`),
                    ),
                );

            await interaction.reply({
                components: [container],
                flags: MessageFlags.IsComponentsV2,
                allowedMentions: { parse: [] },
            });
        }
    }
}

File: apps/bot/src/commands/informations/account.rs
use crate::commands::prelude::*;
use orion_api::OrionError;

/// View your Orion account.
pub struct AccountCommand;

command_meta! {
    META = CommandMeta::builder("account", t!("commands.account.description"))
        .description_localizations(localize_key(|locale| t!(locale, "commands.account.description")))
        .category("Informations")
        .installations(&[
            ApplicationIntegrationType::GuildInstall,
            ApplicationIntegrationType::UserInstall,
        ])
        .contexts(&[
            InteractionContextType::Guild,
            InteractionContextType::BotDm,
            InteractionContextType::PrivateChannel,
        ])
}

#[async_trait]
impl Command<App> for AccountCommand {
    fn meta(&self) -> &CommandMeta {
        &META
    }

    #[t_ns(namespace = "commands.account")]
    async fn handle_command(&self, ctx: &CommandContext<'_>) -> CommandResult {
        let user = ctx.app.orion_api.get_user(ctx.user_id().to_string()).await;

        match user {
            Ok(user) => {
                // The user has an account, build the container

                let container = ContainerBuilder::new()
                    .accent_color(Palette::PRIMARY.int)
                    .add_text_display(|d| {
                        d.content(format!(
                            "## {} Orion - {}",
                            Emojis::LOGO,
                            t!(ctx, "card.title.your_account")
                        ))
                    })
                    .add_text_display(|d| {
                        d.content(t!(ctx, "card.content.credits", amount = user.credits))
                    })
                    .add_action_row(|row| {
                        row.add_button(|b| {
                            b.link(Config::DASHBOARD_URL)
                                .label(t!(ctx, "/common.dashboard"))
                        })
                        .add_button(|b| b.link(Config::PANEL_URL).label(t!(ctx, "/common.panel")))
                    });

                ctx.reply(Reply::new().components_v2(container.into()))
                    .await?;
            }
            Err(OrionError::Api { status: 404, .. }) => {
                // 404 means the user does not have an account yet

                let container = ContainerBuilder::new()
                    .accent_color(Palette::PRIMARY.int)
                    .add_text_display(|d| d.content(format!("## {} Orion", Emojis::LOGO)))
                    .add_text_display(|d| d.content(t!(ctx, "no_account.content.guide")))
                    .add_action_row(|row| {
                        row.add_button(|b| {
                            b.link(Config::DASHBOARD_URL)
                                .label(t!(ctx, "no_account.buttons.create_account"))
                        })
                    });

                ctx.reply(Reply::new().components_v2(container.into()))
                    .await?;
            }
            Err(e) => {
                return Err(anyhow::anyhow!(e).into());
            }
        }

        Ok(())
    }
}
